src/battery-check/battery-check.c

```diff
@@ -0,0 +1,141 @@
+                return;
+        }
+
+        w = strlen(plymouth_message);
+        if (write(fd, plymouth_message, strlen(plymouth_message)) != w)
+                log_full_errno(IN_SET(errno, EAGAIN, ENOENT) || ERRNO_IS_DISCONNECT(errno) ? LOG_DEBUG : LOG_WARNING, errno, "Failed to write to plymouth: %m");
+        w = strlen(plymouth_shutdown);
+        if (write(fd, plymouth_shutdown, strlen(plymouth_shutdown)) != w)
+                log_full_errno(IN_SET(errno, EAGAIN, ENOENT) || ERRNO_IS_DISCONNECT(errno) ? LOG_DEBUG : LOG_WARNING, errno, "Failed to write to plymouth: %m");
+
+}
+
+static int parse_argv(int argc, char * argv[]) {
+
+        enum {
+                ARG_VERSION = 0x100,
+        };
+
+        static const struct option options[] = {
+                { "help",    no_argument, NULL, 'h'         },
+                { "version", no_argument, NULL, ARG_VERSION },
+                {}
+        };
+
+        int c;
+
+        assert(argc >= 0);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "h", options, NULL)) >= 0)
+
+                switch (c) {
+
+                case 'h':
+                        help();
+                        return 0;
+
+                case ARG_VERSION:
+                        return version();
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        if (optind < argc)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "%s takes no argument.",
+                                       program_invocation_short_name);
+        return 1;
+}
+
+static int run(int argc, char *argv[]) {
+        int r;
+
+        log_parse_environment();
+        log_open();
+
+        r = parse_argv(argc, argv);
+        if (r <= 0)
+                return r;
+
+        r = battery_is_discharging_and_low();
+        if (r < 0) {
+                log_warning_errno(r, "Failed to check battery status, ignoring: %m");
+                return 0;
+        }
+        if (r > 0) {
+                _cleanup_close_ int fd = -EBADF;
+                _cleanup_free_ char * message = NULL;
+                log_emergency("Battery level critically low. Powering off in 10 seconds.");
+                fd = open_terminal("/dev/console", O_WRONLY|O_NOCTTY|O_CLOEXEC);
+                if (fd < 0)
+                        log_warning_errno(fd, "Failed to open console, ignoring: %m");
+                else
+                        dprintf(fd, "%s" ANSI_HIGHLIGHT_RED " Battery level critically low. Powering off in 10 seconds. " ANSI_NORMAL "%s\n", special_glyph(SPECIAL_GLYPH_LOW_BATTERY), special_glyph(SPECIAL_GLYPH_LOW_BATTERY));
+                if (asprintf(&message, "%s Battery level critically low. Powering off in 10 seconds.", special_glyph(SPECIAL_GLYPH_LOW_BATTERY)) < 0)
```

Comment:
move this inside the battery_check_send_plymouth_message function too, no point in having one in and one outside

Reply:
well, or maybe better just format this once, and use it both on the console and for the plymouth stuff. i.e.:

```c
if (asprintf(&message, "%s Battery level critically low. Powering off in 10 seconds.", special_glyph(SPECIAL_GLYPH_LOW_BATTERY)) < 0)
        return log_oom();

fd = open_terminal("/dev/console", O_WRONLY|O_NOCTTY|O_CLOEXEC);
if (fd < 0)
        log_warning_errno(fd, …);
else
        dprintf(fd, ANSI_HIGHLIGHT_RED "%s" ANSI_NORMAL "\n", message;

battery_check_send_plymouth_message(message);
```

that would certainly be nicer, since we know both the console and ply will get the exact same message